pkg/v1/cmd.go

```diff
@@ -0,0 +1,620 @@
+	{
+		mergeBaseRaw, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+			"git", "merge-base", "main", "FETCH_HEAD",
+		), dir))
+		if err != nil {
+			return nil, err
+		}
+		mergeBase = strings.TrimSpace(mergeBaseRaw)
+	}
+
+	var downstreamCommits []internal.Commit
+	{
+		rawCommits, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+			"git", "log", mergeBase+"..main",
+			"--ancestry-path", mergeBase,
+			"--no-merges", "--reverse", "--quiet",
+			internal.PrettyFormat,
+		), dir))
+		if err != nil {
+			return nil, err
+		}
+		for _, line := range strings.Split(rawCommits, "\n") {
+			line = strings.TrimSpace(line)
+			if line == "" {
+				continue
+			}
+			info, err := internal.ParseFormat(line)
+			if err != nil {
+				return nil, err
+			}
+			logger = logger.WithFields(logrus.Fields{
+				"commit":  info.Hash,
+				"message": info.Message,
+			})
+			messageMatches := upstreamCommitRegex.FindStringSubmatch(info.Message)
+			if len(messageMatches) == 0 || len(messageMatches[0]) == 0 {
+				return nil, fmt.Errorf("unexpected commit message: %s", info.Message)
+			}
+
+			// TODO: handle reverts, what else?
+			match := strings.Trim(messageMatches[4], "<>:")
+			switch match {
+			case "drop":
+				logger.Info("dropping commit")
+				continue
+			case "carry":
+				logger.Info("carrying commit")
+				downstreamCommits = append(downstreamCommits, info)
+			default:
+				logger.Info("investigating cherry-picked PR")
+				// The UPSTREAM: 1234: format only tells us the upstream pull request that was cherry-picked. Unfortunately,
+				// there's no great way to figure out if that pull request is still something we need to carry or if it's part
+				// of the newer version we're pulling in. The GitHub synthetic ref pull/1234/head gives us the commits that
+				// make up the pull request, but merge strategies other than "merge" will edit those commits - all the v1
+				// repos use such a strategy. We could reach out to the GitHub API, but that's expensive and time consming.
+
+				// Instead, we know that (today) all the v1 repos we're managing use a strategy that renames the commit
+				// to be sufffixed with the pull request "(#1234)", so we can check to see if such a commit exists upstream.
+				// This is not bullet-proof, but the failure mode is that we will try to cherry-pick a commit that does not
+				// need to be there and will either apply as a benign empty commit or fail to apply - both acceptable outcomes.
+				// We won't silently do the wrong thing.
+
+				rawMatches, err := internal.RunCommand(logger, internal.WithDir(exec.CommandContext(ctx,
+					"git", "log", "--pretty=format:%H", "--grep", fmt.Sprintf("(#%s)", match), commit,
+				), dir))
+				if err != nil {
+					return nil, err
+				}
+
+				if len(strings.TrimSpace(rawMatches)) == 0 {
+					logger.Info("cherry-picked PR needs to be carried")
+					downstreamCommits = append(downstreamCommits, info)
```

Comment:
This only gets us the last commit from the PR - the one with (#<pr num>) at the end. Do we lose any other commits from the same PR?

Reply:
Potentially an argument for switching to using merge commits upstream?

Reply:
This appends the commits we had in the repo before, without changing them - so it's up to the person doing the cherry-pick in the first place to get that right, but we won't lose anything